### PR TITLE
feat(vmclass): first implementation

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
first implementation vmclass

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/vmclass/vmclass_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	err = r.updateVMClass(ctx, class)
+	if err != nil {
+		r.logger.Error("Failed to update VirtualMachineClass", slog.String("namespacedName", req.String()))
+		return reconcile.Result{}, err
+	}
+	r.logger.Info("Finished reconcile VM", slog.String("namespacedName", req.String()))
+	return result, nil
+}
+
+func (r *Reconciler) updateVMClass(ctx context.Context, vmClass *service.Resource[*virtv2.VirtualMachineClass, virtv2.VirtualMachineClassStatus]) error {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmclass/vmclass_reconciler.go:137]: Why not just call vmClass.Update(ctx) ? Why do we need this method?

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmclass/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmclasscondition"
+)
+
+const nameDiscoveryHandler = "DiscoveryHandler"
+
+func NewDiscoveryHandler(client client.Client) *DiscoveryHandler {
+	return &DiscoveryHandler{
+		client: client,
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:41]: unused

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	var matchLabels map[string]string
+	var filter filterFunc
+
+	switch curr.Spec.CPU.Type {
+	case virtv2.CPUTypeHost, virtv2.CPUTypeHostPassthrough:
+		return nil, nil
+	case virtv2.CPUTypeDiscovery:
+		matchLabels = curr.Spec.CPU.Discovery.MatchLabels
+		filter = func(nodes []corev1.Node) []corev1.Node {
+			var filtred []corev1.Node
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go:87]: typo: filtered

File: api/core/v1alpha2/virtual_machine_class.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	Memory *SizingPolicyMemory `json:"memory,omitempty"`
+	// Allowed values of the `coreFraction` parameter.
+	CoreFractions []CoreFractionValue `json:"coreFractions,omitempty"`
+	// Allowed values of the `dedicatedCores` parameter.
+	DedicatedCores []bool `json:"dedicatedCores,omitempty"`
+	// The policy applies for a specified range of the number of CPU cores.
+	Cores *SizingPolicyCores `json:"cores,omitempty"`
+}
+
+// TODO: add logic
```
Comment (Isteb4k) [inline api/core/v1alpha2/virtual_machine_class.go:113]: Is this still todo?

File: api/core/v1alpha2/vmclasscondition/condition.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type Reason string
+
+func (r Reason) String() string {
+	return string(r)
+}
+
+const (
+	ReasonNoCpuFeaturesEnabled Reason = "NoCpuFeaturesEnabled"
```
Comment (Isteb4k) [inline api/core/v1alpha2/vmclasscondition/condition.go:37]: Add comments to document types and reasons

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: crds/virtualmachineclasses.yaml [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+                        description: |-
+                          matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+                          map is equivalent to an element of matchExpressions, whose key field is "key", the
+                          operator is "In", and the values array contains only "value". The requirements are ANDed.
+                        type: object
+                    type: object
+                    x-kubernetes-map-type: atomic
+                  features:
+                    description: Required instructions for the CPU as a list More
+                      information about features [here](https://libvirt.org/formatdomain.html#cpu-model-and-topology)
```
Comment (diafour) [inline crds/virtualmachineclasses.yaml:107]: "required" is ambiguous here.

```suggestion
                    description: A list of CPU instructions (features) required for virtual machines of this class.
                      More information about features [here](https://libvirt.org/formatdomain.html#cpu-model-and-topology)
```

File: api/core/v1alpha2/virtual_machine_class.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	// The policy applies for a specified range of the number of CPU cores.
+	Cores *SizingPolicyCores `json:"cores,omitempty"`
+}
+
+// +kubebuilder:validation:Enum={5,10,20,50,100}
+type CoreFractionValue int
+
+type SizingPolicyMemory struct {
+	MemoryMinMax `json:",inline"`
+	// Increase step of memory size from min to max.
```
Comment (diafour) [inline api/core/v1alpha2/virtual_machine_class.go:118]:
```suggestion
	// Memory size discretization step. I.e. min=2Gi, max=4Gi, step=1Gi allows to set virtual machine memory size to 2Gi, 3Gi, or 4Gi.
```

File: api/core/v1alpha2/virtual_machine_class.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:example=1
+	Min int `json:"min"`
+	// Maximum cpu core count.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum=1024
+	// +kubebuilder:example=10
+	Max int `json:"max"`
+	// Increase step of cpu core count from min to max.
```
Comment (diafour) [inline api/core/v1alpha2/virtual_machine_class.go:157]:
```suggestion
	// Cpu cores count discretization step. I.e. min=2, max=10, step=4 allows to set virtual machine cpu cores to 2, 6, or 10.
```

File: api/scripts/update-codegen.sh [hunk 9]
```diff
@@ -19,9 +19,9 @@ set -o pipefail
 
 function usage {
   cat <<EOF
-Usage: $(basename "$0") <core/subresources/all>
+Usage: $(basename "$0") <core/subresources/crds/all>
```
Comment (diafour) [inline api/scripts/update-codegen.sh:22]:
```suggestion
Usage: $(basename "$0") { core | subresources | crds | all }
```

File: api/scripts/update-codegen.sh [hunk 10]
```diff
@@ -30,6 +30,9 @@ function source::settings {
   SCRIPT_ROOT="${SCRIPT_DIR}/.."
   CODEGEN_PKG="$(go env GOMODCACHE)/$(go list -f '{{.Path}}@{{.Version}}' -m k8s.io/code-generator)"
   MODULE="github.com/deckhouse/virtualization/api"
+  PREFIX_GROUP="virtualization.deckhouse.io_"
+  # TODO: legacy, delete after all crds are auto-generated
+  ALLOWED_RESOURCE_GEN_CRD=("VirtualMachineClass ExampleKind1 ExampleKind2")
```
Comment (diafour) [inline api/scripts/update-codegen.sh:35]: New code is not legacy ;)

```suggestion
  # TODO: Temporary filter until all CRDs become auto-generated.
  ALLOWED_RESOURCE_GEN_CRD=("VirtualMachineClass ExampleKind1 ExampleKind2")
```

File: api/scripts/update-codegen.sh [hunk 11]
```diff
@@ -72,6 +75,24 @@ function generate::core {
...
+          cd "${SCRIPT_ROOT}/.."
+          OUTPUT_BASE=$(mktemp -d)
+          trap 'rm -rf "${OUTPUT_BASE}"' ERR EXIT
+
+          "${GOPATH}/bin/controller-gen" crd paths=./api/core/v1alpha2/... output:crd:dir="${OUTPUT_BASE}"
+
+          while IFS= read -r -d '' file
+            do
+              # TODO: legacy, delete after all crds are auto-generated
+              if ! [[ " ${ALLOWED_RESOURCE_GEN_CRD[*]} " =~ [[:space:]]$(yq '.spec.names.kind' "$file")[[:space:]] ]]; then
```
Comment (diafour) [inline api/scripts/update-codegen.sh]: You declare ALLOWED_RESOURCE_GEN_CRD as an array of strings, but there is only one item with multiple kinds. And here you convert array to one string. ALLOWED_RESOURCE_GEN_CRD can be just a string with space separated kinds.

File: api/scripts/update-codegen.sh [hunk 12]
```diff
@@ -72,6 +75,24 @@ function generate::core {
...
+                continue
+              fi
+              cp "$file" "./crds/${file/#"$OUTPUT_BASE/$PREFIX_GROUP"/}"
+          done <   <(find "${OUTPUT_BASE}"/* -type f -iname "*.yaml" -print0)
```
Comment (diafour) [inline api/scripts/update-codegen.sh]:
```
while IFS= read -d ''... 
   ...
done < <(command)
```

is overcomplicated here. Just use for:

```
for file in $(command) ; do
 ....
```

File: api/scripts/update-codegen.sh [hunk 13]
```diff
@@ -72,6 +75,24 @@ function generate::core {
           cp -R "${OUTPUT_BASE}/${MODULE}/." "${SCRIPT_ROOT}"
 }
 
+function generate:crds() {
```
Comment (diafour) [inline api/scripts/update-codegen.sh:78]:
```suggestion
function generate::crds() {
```

for consistency

File: crds/doc-ru-virtualmachine.yaml [hunk 14]
```diff
@@ -367,6 +367,9 @@ spec:
                     * BIOS — использовать BIOS.
                     * EFI — использовать Unified Extensible Firmware (EFI/UEFI).
                     * EFIWithSecureBoot — использовать UEFI/EFI с поддержкой SecureBoot.
+                virtualMachineClassName:
+                  description: |
+                    Имя `virtualMachineClass` ресурс описывающий требования к виртуальному CPU и политику размещения ресурсов.
```
Comment (diafour) [inline crds/doc-ru-virtualmachine.yaml:372]:
```suggestion
                    Имя ресурса `VirtualMachineClass`, который описывает требования к виртуальному CPU, памяти и политику размещения ресурсов.
```

File: crds/virtualmachine.yaml [hunk 15]
```diff
@@ -856,7 +856,10 @@ spec:
                     * BIOS - use legacy BIOS.
                     * EFI - use Unified Extensible Firmware (EFI/UEFI).
                     * EFIWithSecureBoot - use UEFI/EFI with SecureBoot support.
-
+                virtualMachineClassName:
+                  type: string
+                  description: |
+                    Name of `virtualMachineClass` resource describing the requirements for a virtual CPU and the resource allocation policy.
```
Comment (diafour) [inline crds/virtualmachine.yaml:862]:
```suggestion
                    Name of the `VirtualMachineClass` resource describing the requirements for a virtual CPU, memory and the resource allocation policy.
```

File: images/virtualization-artifact/pkg/controller/vmchange/comparators.go [hunk 16]
```diff
@@ -30,6 +32,15 @@ const (
 	DefaultTerminationGracePeriodSeconds = int64(60)
 )
 
+func compareVirtualmachineClass(current, desired *v1alpha2.VirtualMachineSpec) []FieldChange {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmchange/comparators.go:35]:
```suggestion
func compareVirtualMachineClass(current, desired *v1alpha2.VirtualMachineSpec) []FieldChange {
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/deletion.go [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/common"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmclass/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const nameDeletionHandler = "DeletionHandler"
+
+func NewProtectionHandler(client client.Client, recorder record.EventRecorder, logger *slog.Logger) *DeletionHandler {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/deletion.go:38]: Is it deletion or protection?

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	t := current.Spec.CPU.Type
+	mgr := conditions.NewManager(changed.Status.Conditions)
+	if t == virtv2.CPUTypeHostPassthrough || t == virtv2.CPUTypeHost {
+		mgr.Update(conditions.NewConditionBuilder2(vmclasscondition.TypeDiscovered).
+			Generation(current.GetGeneration()).
+			Message(fmt.Sprintf("Discovery available only cpu.type %q", virtv2.CPUTypeDiscovery)).
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:64]:
```suggestion
			Message(fmt.Sprintf("Discovery not needed for cpu.type %q", t)).
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+	if updated := addAllUnknown(changed, vmclasscondition.TypeDiscovered.String()); updated {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	t := current.Spec.CPU.Type
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:59]: I don't like `t` outside of tests.

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+	case virtv2.CPUTypeDiscovery:
+		if len(featuresEnabled) > 0 {
+			cb.Message("").Reason2(vmclasscondition.ReasonDiscoverySucceeded).Status(metav1.ConditionTrue)
+			break
+		}
+		cb.Message("Detection nodes have no features.").
+			Reason2(vmclasscondition.ReasonDiscoveryFailed).
+			Status(metav1.ConditionFalse)
+	default:
+		cb.Message(fmt.Sprintf("Discovery available only cpu.type %q", virtv2.CPUTypeDiscovery)).
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:112]:
```suggestion
		cb.Message(fmt.Sprintf("Discovery not needed for cpu.type %q", t)).
```

Message(fmt.Sprintf("Discovery not needed for cpu.type %q", t)).

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+		}
+	}
+	cb := conditions.NewConditionBuilder2(vmclasscondition.TypeDiscovered).Generation(current.GetGeneration())
+	switch t {
+	case virtv2.CPUTypeDiscovery:
+		if len(featuresEnabled) > 0 {
+			cb.Message("").Reason2(vmclasscondition.ReasonDiscoverySucceeded).Status(metav1.ConditionTrue)
+			break
+		}
+		cb.Message("Detection nodes have no features.").
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:108]:
```suggestion
		cb.Message("No common features are discovered on nodes.").
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+	if s.vmClass == nil || s.vmClass.IsEmpty() {
+		return nil, nil
+	}
+	name := s.vmClass.Current().GetName()
+	vms := &virtv2.VirtualMachineList{}
+	err := s.client.List(ctx, vms, client.MatchingFields{
+		indexer.IndexFieldVMByClass: name,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list virtual machine: %w", err)
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go:65]:
```suggestion
		return nil, fmt.Errorf("failed to list virtual machines by vmclass %s: %w", name, err)
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go (hunk 3 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go:87]: Still `filtred`

File: images/virtualization-artifact/pkg/controller/vmclass/internal/deletion.go (hunk 17 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vmclass/internal/deletion.go:38]: fixed

File: images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go (hunk 19 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go:59]: fixed

File: api/scripts/update-codegen.sh (hunk 12 above)
Comment (yaroslavborbat) [reply api/scripts/update-codegen.sh]: ok

File: api/scripts/update-codegen.sh (hunk 11 above)
Comment (yaroslavborbat) [reply api/scripts/update-codegen.sh]: fixed

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go (hunk 3 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go]: fixed

**Review by Isteb4k — APPROVED** [review]

**Review by Isteb4k — APPROVED** [review]

**Pull request merged**